File: windows/src/Restomenum.Agent.Core/Backoff.cs
namespace Restomenum.Agent.Core;

/// <summary>
/// Yeniden bağlanma gecikmesi (§5.2) — **jitter ZORUNLUDUR**.
///
/// Bir gateway instance'ı kapandığında ona bağlı tüm agent'lar aynı anda kopar. Sabit gecikmeyle
/// hepsi aynı saniyede geri dönerse **yeniden bağlanma fırtınası** olur: 10.000 cihazda ~33
/// bağlantı/sn, 50.000'de ~166. Jitter bu dalgayı yayar.
///
/// Jitter **çift yönlüdür** (±%20). Yalnız pozitif jitter dalgayı yaymak yerine öteler.
/// </summary>
public sealed class Backoff
{
    private static readonly long[] StepsMs = { 500, 1_000, 2_000, 5_000, 10_000, 30_000 };
    private readonly Func<double, double> _random;
    private int _attempt;

    /// <param name="random">[0,x) aralığında değer üretir. Testte sabitlenebilir.</param>
    public Backoff(Func<double, double>? random = null) =>
        _random = random ?? (x => Random.Shared.NextDouble() * x);

    /// <summary>Başarılı bağlantıdan sonra çağrılır — sonraki kopma en kısa gecikmeyle başlasın.</summary>
    public void Reset() => _attempt = 0;

    public long Next()
    {
        var baseMs = StepsMs[Math.Min(_attempt, StepsMs.Length - 1)];
        _attempt++;
        var spread = baseMs * 0.2;
        var delta = _random(2 * spread) - spread;
        return Math.Max(1, (long)(baseMs + delta));
    }

    public int AttemptCount => _attempt;
}

File: android/core/src/main/kotlin/com/restomenum/agent/Backoff.kt
package com.restomenum.agent

import kotlin.random.Random

/**
 * Yeniden bağlanma gecikmesi (§5.2) — **jitter ZORUNLUDUR**.
 *
 * Bir gateway instance'ı kapandığında ona bağlı tüm agent'lar aynı anda kopar. Sabit gecikmeyle
 * hepsi aynı saniyede geri dönerse **yeniden bağlanma fırtınası** olur: 10.000 cihazda ~33
 * bağlantı/sn, 50.000'de ~166. Jitter bu dalgayı yayar.
 *
 * Basamaklar: 0.5s, 1s, 2s, 5s, 10s, maks 30s (±%20 jitter).
 */
class Backoff(private val rastgele: (Double) -> Double = { Random.nextDouble(it) }) {
    private val basamaklarMs = longArrayOf(500, 1_000, 2_000, 5_000, 10_000, 30_000)
    private var deneme = 0

    /** Başarılı bağlantıdan sonra çağrılır — bir sonraki kopma en kısa gecikmeyle başlasın. */
    fun sifirla() { deneme = 0 }

    /** Sıradaki gecikme (ms), jitter uygulanmış. */
    fun sonraki(): Long {
        val taban = basamaklarMs[minOf(deneme, basamaklarMs.size - 1)]
        deneme++
        // ±%20 jitter: dalgayı yayar. Yalnız pozitif jitter, dalgayı yaymak yerine ötelerdi.
        val sapma = taban * 0.2
        val delta = rastgele(2 * sapma) - sapma
        return (taban + delta).toLong().coerceAtLeast(1)
    }

    fun denemeSayisi(): Int = deneme
}
